pantheon: manage user session with systemd

bobby285271 opened the pull request:
###### Motivation for this change

Fixes #101817

- https://github.com/elementary/session-settings/pull/55
- https://github.com/elementary/gala/pull/1255
- https://github.com/elementary/gala/pull/1256

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

davidak commented:
Is this ready to test? Asking because it's marked as draft.

bobby285271 commented:
> Is this ready to test?

Yes, I am using this on my physical machine.

bobby285271 commented on pkgs/desktops/pantheon/desktop/gala/default.nix:
```diff
@@ -78,25 +78,6 @@ stdenv.mkDerivation rec {
-    # Hope this can make WITH_SYSTEMD block functions
-    # and make libsystemd found by pkg-config
-    # https://github.com/NixOS/nixpkgs/issues/101817#issuecomment-932956940
-    ./fix-gala-x11-timeout.patch
```
The patch is accepted on upstream, however seems that I cannot change this part to `fetchpatch` as gala fail to build.

- https://github.com/NixOS/nixpkgs/issues/32084

Maybe I will just bump gala to a unstable version, there are only translation updates beside the systemd unit change.

bobby285271 commented on pkgs/desktops/pantheon/desktop/gala/default.nix:
```diff
@@ -71,12 +73,34 @@ stdenv.mkDerivation rec {
+      url = "https://github.com/elementary/gala/commit/bd362ec0c89bd72c5afca3f8313ee70d7ee25357.patch";
+      sha256 = "1l8dzyk7yncppg8yzhk7ddpccg3c0j95sar19hckc4wy4c1bn1j0";
+    })
+    # Hope this can make WITH_SYSTEMD block functions
+    # and make libsystemd found by pkg-config
+    # https://github.com/NixOS/nixpkgs/issues/101817#issuecomment-932956940
+    ./fix-gala-x11-timeout.patch
```
The patch is accepted on upstream, however seems that I cannot change this part to `fetchpatch` as gala fail to build.

```nix
(fetchpatch {
  url = "https://github.com/elementary/gala/commit/cc8692ab2766d3fde42ebf8eaadeab3f5b3c45f7.patch";
  sha256 = "0c2nwawfdikqfcas4mffcp7klfl9wxxr3m5b7pxwl6iyycg4mwg2";
})
```

Related issue:

- https://github.com/NixOS/nixpkgs/issues/32084

bobby285271 commented:
@ofborg test pantheon

davidak commented on nixos/modules/services/x11/desktop-managers/pantheon.nix:
```diff
@@ -168,8 +168,11 @@ in
-      systemd.packages = [
-        pkgs.gnome.gnome-settings-daemon338
+      systemd.packages = with pkgs; [
+        gnome.gnome-settings-daemon338
```
does it make sense to have both?

davidak commented on pkgs/desktops/pantheon/desktop/elementary-session-settings/default.nix:
```diff
@@ -71,7 +72,7 @@ let
     export XDG_DATA_DIRS=@out@/share:$XDG_DATA_DIRS
 
     # Start pantheon session. Keep in sync with upstream
```
maybe add a link where upstream has it defined?

davidak commented:
This seem to work.

Tested:

- booting several times
- login
- lock screen
- suspend, resume
- start applications

davidak reviewed (approved):
The changes look good to me and work.

bobby285271 commented:
I updated the PR description. I guess I should close this for now to avoid confusions, I don't plan to merge this for now. I will re-open this or reflect this changes in other PR when ready.

davidak commented:
I noticed today that i can't always click, for example moving windows or make a different video active by clicking on it don't work. Also, hotkeys for screenshot does not work anymore. Screenshots using the app work. Nothing related in logs.

bobby285271 commented:
I can reproduce the keyboard keys issue but not other yet. I guess it will be nice to report the issue upstream as I don't know how to deal with these downstream currently. Anyway I am tracking this in https://github.com/NixOS/nixpkgs/projects/21#card-73471036 so we can make sure we revisit this when doing Pantheon 7 update.

davidak commented:
The described problems are actually not related to systemd session. They also occur with latest stable NixOS 21.11.334139.1bd4bbd49be.

bobby285271 commented:
For me, at least the keyboard keys issue happens only with `--builtin` removed...

Filechooser portal should work fine on chromium and for your information the filechooser has been redesigned in Pantheon 6.0.4.

bobby285271 commented on nixos/modules/services/x11/desktop-managers/pantheon.nix:
```diff
@@ -168,12 +168,38 @@ in
       programs.dconf.enable = true;
       networking.networkmanager.enable = mkDefault true;
 
+      systemd.user.targets."gnome-session-initialized".wants = [
+        "org.gnome.SettingsDaemon.A11ySettings.target"
+        "org.gnome.SettingsDaemon.Color.target"
+        "org.gnome.SettingsDaemon.Datetime.target"
+        "org.gnome.SettingsDaemon.Housekeeping.target"
+        "org.gnome.SettingsDaemon.Keyboard.target"
+        "org.gnome.SettingsDaemon.MediaKeys.target"
+        "org.gnome.SettingsDaemon.Power.target"
+        "org.gnome.SettingsDaemon.PrintNotifications.target"
+        "org.gnome.SettingsDaemon.Rfkill.target"
+        "org.gnome.SettingsDaemon.ScreensaverProxy.target"
+        "org.gnome.SettingsDaemon.Sharing.target"
+        "org.gnome.SettingsDaemon.Smartcard.target"
+        "org.gnome.SettingsDaemon.Sound.target"
+        "org.gnome.SettingsDaemon.UsbProtection.target"
+        "org.gnome.SettingsDaemon.Wacom.target"
+        "org.gnome.SettingsDaemon.Wwan.target"
+      ];
+
+      systemd.user.targets."gnome-session-x11-services".wants = [
+        "org.gnome.SettingsDaemon.XSettings.target"
+      ];
+
```
Not sure if we should do this here...? But this should fix the media key issue...

bobby285271 commented on nixos/modules/services/x11/desktop-managers/pantheon.nix:
```diff
@@ -168,12 +168,35 @@ in
       programs.dconf.enable = true;
       networking.networkmanager.enable = mkDefault true;
 
+      systemd.user.targets."gnome-session-initialized".wants = [
+        "org.gnome.SettingsDaemon.A11ySettings.target"
+        "org.gnome.SettingsDaemon.Color.target"
+        "org.gnome.SettingsDaemon.Datetime.target"
+        "org.gnome.SettingsDaemon.Housekeeping.target"
+        "org.gnome.SettingsDaemon.Keyboard.target"
+        "org.gnome.SettingsDaemon.MediaKeys.target"
+        "org.gnome.SettingsDaemon.Power.target"
+        "org.gnome.SettingsDaemon.PrintNotifications.target"
+        "org.gnome.SettingsDaemon.Rfkill.target"
+        "org.gnome.SettingsDaemon.ScreensaverProxy.target"
+        "org.gnome.SettingsDaemon.Sharing.target"
+        "org.gnome.SettingsDaemon.Smartcard.target"
+        "org.gnome.SettingsDaemon.Sound.target"
+        "org.gnome.SettingsDaemon.UsbProtection.target"
+        "org.gnome.SettingsDaemon.Wacom.target"
+        "org.gnome.SettingsDaemon.Wwan.target"
+        "org.gnome.SettingsDaemon.XSettings.target"
+      ];
+
```
Maybe this can fix the media key issue (cc @davidak) but I am not sure if this is the correct place to put this... Shall I update [`nixos/modules/services/desktops/gnome/gnome-settings-daemon.nix`](https://github.com/NixOS/nixpkgs/blob/bb36c044d1be1be5f023fc9f4a5dc38b5c023787/nixos/modules/services/desktops/gnome/gnome-settings-daemon.nix#L60-L80) instead (cc @jtojnar) or shall this included in upstream (cc @davidmhewitt)?

davidmhewitt replied:
It's been a while since I've looked at this, and I don't think I fully understand what you're doing with the nix packaging here, but the way this is supposed to work is that the `gnome-settings-daemon` package installs a set of systemd units and targets that automatically start with `gnome-session` targets.

Since we launch Pantheon with `gnome-session`, all of the GNOME SettingsDaemon plugins should just work without us explicitly having to specify we want them to start, so I don't think there's anything to do "upstream" in elementary here. But please let us know what you find.

Is this potentially an issue where different versions of GNOME components are being used together for some reason?

bobby285271 replied:
@davidmhewitt Haven't tested this, but I found this in `/etc/systemd/user/gnome-session@gnome.target.d/gnome.session.conf`:

```
[Unit]
# Must be in sync with gnome.session
Wants=org.gnome.SettingsDaemon.A11ySettings.target
Wants=org.gnome.SettingsDaemon.Color.target
Wants=org.gnome.SettingsDaemon.Datetime.target
Wants=org.gnome.SettingsDaemon.Housekeeping.target
Wants=org.gnome.SettingsDaemon.Keyboard.target
Wants=org.gnome.SettingsDaemon.MediaKeys.target
Wants=org.gnome.SettingsDaemon.Power.target
Wants=org.gnome.SettingsDaemon.PrintNotifications.target
Wants=org.gnome.SettingsDaemon.Rfkill.target
Wants=org.gnome.SettingsDaemon.ScreensaverProxy.target
Wants=org.gnome.SettingsDaemon.Sharing.target
Wants=org.gnome.SettingsDaemon.Smartcard.target
Wants=org.gnome.SettingsDaemon.Sound.target
Wants=org.gnome.SettingsDaemon.UsbProtection.target
Wants=org.gnome.SettingsDaemon.Wacom.target
Wants=org.gnome.SettingsDaemon.XSettings.target
```

I thought there should have something similar for `gnome-session@pantheon.target`...

bobby285271 replied:
Filed https://github.com/elementary/session-settings/pull/63

bobby285271 commented on nixos/modules/services/x11/desktop-managers/pantheon.nix:
```diff
@@ -168,12 +168,35 @@ in
+        "org.gnome.SettingsDaemon.Sound.target"
+        "org.gnome.SettingsDaemon.UsbProtection.target"
+        "org.gnome.SettingsDaemon.Wacom.target"
+        "org.gnome.SettingsDaemon.Wwan.target"
+        "org.gnome.SettingsDaemon.XSettings.target"
```
Ooops `org.gnome.SettingsDaemon.Wwan.target` is not in `/etc/systemd/user/gnome-session@gnome.target.d/gnome.session.conf`...

And seems that `systemd.packages` isn't picking `.wants` directories...

bobby285271 replied:
Maybe `nixos/modules/services/desktops/gnome/gnome-settings-daemon.nix` needs cleanup, I opened #153840, I think I also need a `services.gnome.gnome-settings-daemon.package` option in the future, maybe I will do that when this PR is ready.

jtojnar replied:
See also https://github.com/NixOS/nixpkgs/commit/68ab37aa4491c8f870483c9eff888dc5fc88d2c8, not sure what the reasoning was any more.

bobby285271 replied:
Ooops, then I guess https://github.com/NixOS/nixpkgs/pull/153840 itself is a reason...

bobby285271 commented:
@ofborg test pantheon

bobby285271 commented:
@ofborg test pantheon

bobby285271 commented:
(Let's see when [gnome-session!93](https://gitlab.gnome.org/GNOME/gnome-session/-/merge_requests/93) will go live :upside_down_face:)

bobby285271 committed:
nixos/display-managers: Don't force graphical-session.target activation for Pantheon

See Nixpkgs pull request 233981.

bobby285271 committed:
pantheon: Manage user session with systemd

This is now needed as gnome-session 46 drops `--builtin` option.

bobby285271 committed:
nixosTests.pantheon: Fix gala environ subtest

This is changed after we manage session with systemd, as
io.elementary.gala@.service has `ExecStart=@bindir@/gala`.

bobby285271 committed:
nixos/pantheon: Add missing gala-daemon systemd services

Otherwise it doesn't autostart after user session managed by systemd.

bobby285271 committed:
nixosTests.pantheon: Add test for io.elementary.gala.daemon@x11.service

Which is manually added in the nixos module.

bobby285271 commented:
https://github.com/elementary/session-settings/pull/73 has been merged, I would like to land this soon
